src/apps/vlan/vlan.lua

```diff
@@ -0,0 +1,174 @@
+module(..., package.seeall)
+
+local packet = require("core.packet")
+local bit = require("bit")
+local ffi = require("ffi")
+
+local C = ffi.C
+local receive, transmit = link.receive, link.transmit
+local cast = ffi.cast
+
+Tagger = {}
+Untagger = {}
+
+-- 802.1q
+local dotq_tpid = 0x8100
+local o_ethernet_ethertype = 12
+local uint32_ptr_t = ffi.typeof('uint32_t*')
+
+
+-- build a VLAN tag consisting of 2 bytes of TPID set to 0x8100 followed by the
+-- TCI field which in turns consists of PCP, DEI and VID (VLAN id). Both PCP
+-- and DEI is always 0
+local function build_tag(vid)
+   return ffi.C.htonl(bit.bor(bit.lshift(dotq_tpid, 16), vid))
+end
+
+-- pop a VLAN tag (4 byte of TPID and TCI) from a packet
+function pop_tag(pkt)
+   local payload = pkt.data + o_ethernet_ethertype
+   local length = pkt.length
+   pkt.length = length - 4
+   C.memmove(payload, payload + 4, length - o_ethernet_ethertype - 4)
+end
+
+-- push a VLAN tag onto a packet
+function push_tag(pkt, tag)
+   local payload = pkt.data + o_ethernet_ethertype
+   local length = pkt.length
+   pkt.length = length + 4
+   C.memmove(payload + 4, payload, length - o_ethernet_ethertype)
+   cast(uint32_ptr_t, payload)[0] = tag
+end
+
+-- extract TCI (2 bytes) from packet, no check is performed to verify that the
+-- packet is carrying a VLAN tag, if it's an untagged frame these bytes will be
+-- Ethernet payload
+function extract_tci(pkt)
+   return ffi.C.ntohs(ffi.cast("uint16_t*", packet.data(pkt) + o_ethernet_ethertype + 2)[0])
+end
+
+-- extract VLAN id from TCI
+function tci_to_vid(tci)
+   return bit.band(tci, 0xFFF)
+end
+
+
+function Tagger:new(conf)
+   local o = setmetatable({}, {__index=Tagger})
+   o.tag = build_tag(assert(conf.tag))
+   return o
+end
+
+function Tagger:push ()
+   local input, output = self.input.input, self.output.output
+   local tag = self.tag
+   for _=1,link.nreadable(input) do
+      local pkt = receive(input)
+      push_tag(pkt, tag)
+      transmit(output, pkt)
+   end
+end
+
+function Untagger:new(conf)
+   local o = setmetatable({}, {__index=Untagger})
+   o.tag = build_tag(assert(conf.tag))
+   return o
+end
+
+function Untagger:push ()
+   local input, output = self.input.input, self.output.output
+   local tag = self.tag
+   for _=1,link.nreadable(input) do
+      local pkt = receive(input)
+      local payload = pkt.data + o_ethernet_ethertype
+      if cast(uint32_ptr_t, payload)[0] ~= tag then
+         -- Incorrect VLAN tag; drop.
+         packet.free(pkt)
+      else
+         pop_tag(pkt)
+         transmit(output, pkt)
+      end
+   end
+end
+
+
+VlanMux = {}
+function VlanMux:new(conf)
+   local self = setmetatable({}, {__index=VlanMux})
+   self.ethertype_8100 = ffi.cast("uint16_t", 0x0081) -- 0x8100 in network byte order
```

Comment:
AFAIK there's no pattern. Some apps call its new argument `arg` or `conf`. What is true is that new methods always receive one argument. If the constructor needs more than one argument, a table is passed. Many apps use this pattern:

```
function App:new(arg)
   local conf = config.parse_app_arg(arg)
```

Now, checking what `parse_app_arg` it converts a string with a pattern like "{ property = value,...}" into a Lua table. I think this is legacy code and it's not longer necessary (I'm not aware of any object that passes a string like a Lua table).